File: sparcs/sparcs_workflow/scripts/chunk_extracted_seqs.py
################################################################################
# Split the file of extracted sequences into chunks
#
# Author: Kobie Kirven
# Assmann and Bevilacqua Labs
# The Penn State University
################################################################################

# -- Imports -- #
import argparse
import os

#######################
#  --  Functions --   #
#######################


def split_file_by_line(filename, n):
    """Split a file into n chunks by line"""
    n = int(n)
    with open(filename, "r") as f:
        chunk_size = sum(1 for line in f) // n
        f.seek(0)
        remainder = sum(1 for line in f) % n
        f.seek(0)
        start = 0
        for i in range(n):
            chunk = []
            for j in range(chunk_size + (i < remainder)):
                chunk.append(f.readline())
            yield chunk


if __name__ == "__main__":

    # Parse the arguments
    parser = argparse.ArgumentParser(description="Break up VCF file")
    parser.add_argument("--input", dest="input", help="input file")
    parser.add_argument("--dir", dest="dir", help="working directory")
    parser.add_argument(
        "--chunk-total", dest="chunk", help="The number of files to chop the input into"
    )
    args = parser.parse_args()

    # Make a directory to store the VCF chunks
    if not os.path.exists(f"{args.dir}/extracted_seqs_chunks/"):
        os.system(f"mkdir -p {args.dir}/extracted_seqs_chunks/")

    # Split the extracted sequences into chunks
    for i, chunk in enumerate(split_file_by_line(args.input, args.chunk)):
        with open(
            f"{args.dir}/extracted_seqs_chunks/extracted_flank_snp_{i+1}.txt", "w"
        ) as f:
            f.writelines(chunk)

File: variantfoldrna/variantfoldrna_workflow/source/chunk_extracted_seqs.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"path/filepath"
)

func splitFileByLine(filename string, n int, nullFlag bool, outputDir string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lineCount := 0
	for scanner.Scan() {
		lineCount++
	}
	if err := scanner.Err(); err != nil {
		return err
	}

	chunkSize := lineCount / n
	remainder := lineCount % n

	file.Seek(0, 0)
	for i := 0; i < n; i++ {
		var chunkFileName string
		if nullFlag {
			chunkFileName = filepath.Join(outputDir, "extracted_sequences_null", fmt.Sprintf("extracted_seqs_null_unique_chunk_%d.txt", i+1))
		} else {
			chunkFileName = filepath.Join(outputDir, "extracted_seqs_chunks", fmt.Sprintf("extracted_flank_snp_%d.txt", i+1))
		}

		chunkFile, err := os.Create(chunkFileName)
		if err != nil {
			return err
		}
		defer chunkFile.Close()

		chunkWriter := bufio.NewWriter(chunkFile)
		linesToWrite := chunkSize
		if i < remainder {
			linesToWrite++
		}

		for j := 0; j < linesToWrite; j++ {
			scanner.Scan()
			chunkWriter.WriteString(scanner.Text() + "\n")
		}
		chunkWriter.Flush()
	}

	return nil
}

func main() {
	inputFile := flag.String("input", "", "input file")
	outputDir := flag.String("dir", "", "working directory")
	nullFlag := flag.Bool("null", false, "null file")
	chunkTotal := flag.Int("chunk-total", 0, "The number of files to chop the input into")
	flag.Parse()

	if *inputFile == "" || *outputDir == "" || *chunkTotal == 0 {
		flag.PrintDefaults()
		os.Exit(1)
	}

	// Make a directory to store the chunks
	chunksDir := filepath.Join(*outputDir, "extracted_seqs_chunks")
	if err := os.MkdirAll(chunksDir, os.ModePerm); err != nil {
		fmt.Println("Error creating chunks directory:", err)
		os.Exit(1)
	}

	nullDir := filepath.Join(*outputDir, "extracted_sequences_null")
	if err := os.MkdirAll(nullDir, os.ModePerm); err != nil {
		fmt.Println("Error creating null directory:", err)
		os.Exit(1)
	}

	if err := splitFileByLine(*inputFile, *chunkTotal, *nullFlag, *outputDir); err != nil {
		fmt.Println("Error splitting file:", err)
		os.Exit(1)
	}

	fmt.Println("File split successfully.")
}
